server/src/main/java/no/fractal/socket/send/AbstractMessage.java

```diff
@@ -0,0 +1,150 @@
+
+import no.fractal.socket.StreamUtil;
+import no.fractal.util.StreamUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+
+import java.io.*;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class AbstractMessage {
+
+    private final String messageType;
+    private final ArrayList<Segment> segments;
+
+
+    protected AbstractMessage(String messageType
+    ) {
+        this.messageType = messageType;
+        this.segments = new ArrayList<>();
+    }
+
+    protected void addSegment(Segment segment){
+        this.segments.add(segment);
+    }
+
+
+    public void writeToStream(BufferedOutputStream outputStream) throws IOException{
+
+        byte[] typeBytes = this.messageType.getBytes();
+        byte[] segmentBytes = getMainSegmentMeta().toString().getBytes();
+
+        // byte count
+        int byteCont = segments.stream().map(Segment::getBodyByteSize).reduce(Integer::sum).get();
+        byteCont += segmentBytes.length;
+        byteCont += typeBytes.length;
+        byteCont += 14; // hederstuff
+
+        // - payload tot size 4B
+        StreamUtils.writeInt(outputStream, byteCont);
+
+        // - id Size 2b and id
+        StreamUtils.writeInt(outputStream, typeBytes.length);
+        outputStream.write(typeBytes);
+
+        // - segmentArr size 2b and arr
+        StreamUtils.writeInt(outputStream, segmentBytes.length);
+        outputStream.write(segmentBytes);
+
+        for (Segment segment: segments){
+            segment.writeToStream(outputStream);
+        }
+    }
+
+    private JSONArray getMainSegmentMeta(){
+        JSONArray ret = new JSONArray();
+        segments.forEach(segment -> ret.put(segment.segmentMeta));
+        return ret;
+    }
+
+    private abstract static class Segment{
+        protected final Map<String,String> segmentMeta;
+
+        protected Segment(Map<String, String> segmentMeta) {
+            this.segmentMeta = segmentMeta;
+        }
+
+        Map<String, String> getSegmentMeta() {
+            return segmentMeta;
+        }
+
+        public int getBodyByteSize(){
+            return Integer.parseInt(segmentMeta.get(keys.size.name()));
+        }
+
+        public abstract void writeToStream(BufferedOutputStream outputStream) throws IOException;
+
+        protected enum keys{
+            size,
+            mime_type,
+            file_name
+        }
+    }
+
+    public static class FileSegment extends Segment{
+        public FileSegment(File file)  {
+            super(new HashMap<>());
+            this.file = file;
+
+            try {
+                this.segmentMeta.put(keys.size.name(), String.valueOf(file.length()));
+                this.segmentMeta.put(keys.mime_type.name(), String.valueOf(Files.probeContentType(file.toPath())));
+                this.segmentMeta.put(keys.file_name.name(), file.getName());
+
+            } catch (Exception e){ e.printStackTrace();}
```

Comment:
Exception bør vel finne veien ut, for å bare drepe hele messagen, logge og mulighet for evt sende feil melding til klient

Reply:
hva skal vi jøre med det tenker du kill eller gi besked. Det som er ambefalt er vel kill men siden vi bare hanskes med påloggede kjente node kan vel en feilmelding sendes? 
jeg er usikker skal vi lage er feilmeldingsystem eller skal vi bare anta at alt virker?